minvws/nl-kat-coordination
# Combined schedulers

### Changes

* `app.py`: Removal of creation of organisational schedulers. We don't gather all available organisations from the katalogus and create individual boefje, report, and normalizer schedulers.
* `app.py`: Removal of continuous checking of added or removed organisations. We removed the `monitor_organisations` method that was running in a thread to check if we added or removed organisation and create schedulers for it.
* `models/` : `Task`  and `Schedule` get a specific `organisation` field. Since we don't differentiate tasks that are pushed on an organisation queue anymore, we need to be able to determine for what organisation an item on a queue is from.[^1]
* `schedulers/`: Since we've consolidated the organisational schedulers to 3 types of schedulers, and we assume that we get task creation from consolidated message queues, we need to update the schedulers to handle that.
  * `BoefjeScheduler`
    * retrieve scan profile mutations from one message queue
    * iterate over organisation to check for enables boefjes[^2]
    * refactor of method names and exception handling
* `server/` 
  * Merger of `queues/` and `schedulers/` endpoint. The notion of a queue and scheduler are used interchangeably. Therefore it makes more sense at the moment to merge those endpoints to avoid confusion and simplify the code.
  * Support to pop off more than one `Task` from a scheduler. Task runner will now be able to batch pop tasks from a scheduler using filters.

### Next steps and impact

* Pop endpoint has changed from `/queues` to `/schedulers/{id}/pop` , additionally it will return a paginated result instead of a single `Task` , this is because the pop endpoint now supports filtering with multiple tasks returns. Services that rely on the scheduler pop endpoint need to update their interfaces (https://github.com/minvws/nl-kat-coordination/issues/3961)
* Push endpoint changed from `/queues` to `schedulers/{id}/push`, services that interface with the push endpoints (rocky) need to update their interfaces. (https://github.com/minvws/nl-kat-coordination/issues/3962)
* scan profile mutation message queue, for every organisation a message queue is created for scan profile mutations, this needs to be updated and all scan profile mutations for every organisation needs to be relayed on a single scan profile mutations message queue (https://github.com/minvws/nl-kat-coordination/issues/3963)
* raw data file received message queue (the as with scan profile mutations) (https://github.com/minvws/nl-kat-coordination/issues/3964)
* Model definitions updates: `organisation` fields are added to `Task` , `Schedule` , services using these models need to update their specifications. (https://github.com/minvws/nl-kat-coordination/issues/3965)
* Batch status updates. Several places in the scheduler we can consider batch updating status field of task. Potentially exposing this as a endpoint. This because the task runner will be able to pop off multiple task from a scheduler, it might therefore be beneficial to batch update the tasks. (
* Deleting of organisations needs to be addressed, and what the protocol needs to be with regards of queued tasks, and schedules. (https://github.com/minvws/nl-kat-coordination/issues/3966)

### Issue link

https://github.com/minvws/nl-kat-coordination/issues/3838

### QA notes

The process of scanning tasks, rescheduling of tasks should remain the same from a users standpoint of view. Testing this would be the main focus point. Additionally since all organisational schedulers have been combined to one, particular emphasis should be made to check if users will only see, or are able to interact with tasks that are for the organisation that they are allowed to.

It would be interesting to see if with high volumes the execution of tasks will be done as expected by the user.

---

### Code Checklist







---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

[^1]: The attentive reader will know that the organisation information is present in the `data` specification in a `Task` , and can be filtered on by using a `FilterRequest`. However in practice the scheduler is mainly used in the context of multiple organisations and filtering on this id is deemed developer friendly when interfacing with the scheduler. Hence the choice explicitly defining the `organisation` field on a top level.
[^2]: as mentioned in the code this iterating over all organisations in order to check if there are enabled plugins is very inefficient, please refer to issue #3357
